### PR TITLE
Implement MuSig on Ristretto

### LINKED ISSUE #58: Provide Schnorr Signatures
Provide a easy-to-use API for Schnorr Signatures, independent of underlying EC

### LINKED ISSUE #59: Partial signatures
Provide a easy-to-use API for partial signature generation, verifying and aggregating

### LINKED ISSUE #62: Aggregated signature API
provide a general, easy to use API for building aggregated signatures

Provide one Implementation using Schnorr signatures

### LINKED ISSUE #63: Implement MuSig
Implement MuSig as an aggregated signature scheme. 

Requires #64, requires #62

### PR DESCRIPTION
**Pull request by CjS77**
## Description
Implement MuSig as described in [the Blockstream paper](https://eprint.iacr.org/2018/068)

This implementation is implemented on two layers: A public API layer (`MuSig`), and a private Finite State Machine. The purpose of doing it like this has a few advantages from a security perspective:
* All private data is completely inaccessible to users
* Sensitive temporary data is dropped as the state machine progresses.
* Any invalid actions immediately causes the `MuSig` state to become invalid, and all sensitive data is immediately destroyed.

The state transition set is highly constrained. As a result it is impossible to rewind the MuSig ceremony in order to execute a replay attack; The Message that is being signed can be set exactly once.

`MuSig` takes ownership of the public nonces, mitigating the risk of exposing private keys due to nonce reuse (it's not impossible, but a user must explicitly `clone()` a public nonce, which should give him a hint that he's doing something dangerous).

### Breaking changes
There are a few changes to the API, but the impact is minimal:

* `SecretKeyFactory` is folded into the `SecretKey` trait
* `SchnorrSignature::verify` now takes a `SecretKey` reference rather than a challenge. The original `verify` function has been renamed `verify_challenge`. This allows us to cache the challenge in MuSig internally rather than having to rehash it every time an aggregated signature check is requested.

## Motivation and Context
MuSig provides compact signature aggregation using Schnorr signatures and is useful in many n-of-n multi-signature applications.

Closes #63 
Closes #58 
Closes #59
Closes #62

## How Has This Been Tested?
~20 unit tests are provided

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [x] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [x] Documentation

## Checklist:

* [x] I'm merging against the `development` branch
* [x] I ran `cargo-fmt --all` before pushing
* [x] My change requires a change to the documentation.
* [x] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

### CODE REVIEW

**Comment by philipr-za** [comment]
Oh, there are also several very trivial to fix warnings on this PR.

File: infrastructure/crypto/src/musig.rs [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+        }
+    }
+
+    /// Returns true if every item in the set has been set. An empty set returns true as well.
+    pub fn is_full(&self) -> bool {
+        self.items.iter().all(|v| v.is_some())
+    }
+
+    /// Return the index of the given item in the set by performing a linear search through the set
+    pub fn slow_search(&self, val: &T) -> Option<usize> {
```
[the remark this answers is not in the document]
Comment (CjS77) [reply infrastructure/crypto/src/musig.rs]: Cool, yeah.

File: infrastructure/crypto/src/ristretto/musig.rs [hunk 2]
```diff
@@ -0,0 +1,960 @@
...
+        let joint_key = match &self.state {
+            MuSigState::NonceHashCollection(s) => &s.joint_key,
+            MuSigState::NonceCollection(s) => &s.joint_key,
+            MuSigState::SignatureCollection(s) => &s.joint_key,
+            _ => return None,
+        };
+        joint_key.index_of(pub_key).ok()
+    }
+
+    pub fn add_public_key(self, key: &RistrettoPublicKey) -> Self {
```
[the remark this answers is not in the document]
Comment (CjS77) [reply infrastructure/crypto/src/ristretto/musig.rs]: Sure. I can probably add a bit more detail to the docs up top as well.

File: infrastructure/crypto/src/ristretto/musig.rs [hunk 3]
```diff
@@ -0,0 +1,960 @@
...
+        let key = key.clone();
+        self.handle_event(MuSigEvent::AddKey(key))
+    }
+
+    pub fn set_message(self, msg: &MessageHash) -> Self {
```
[the remark this answers is not in the document]
Comment (CjS77) [reply infrastructure/crypto/src/ristretto/musig.rs]: I'm agnostic on this point. We could accept an arbitrary byte string as the message. I had some concern about a resource attack if an enormous message is passed in, but I suppose we could easily hash it right there and discard the source (or more likely, take a reference)

File: infrastructure/crypto/src/challenge.rs [hunk 4]
```diff
@@ -23,6 +23,7 @@
 use digest::{generic_array::typenum::U32, FixedOutput};
 use sha2::Digest;
 
+pub type MessageHash = Vec<u8>;
```
[the remark this answers is not in the document]
Comment (CjS77) [reply infrastructure/crypto/src/challenge.rs]: Yeah, not a bad idea. I'd check it in the MuSig constructor though.

File: infrastructure/crypto/src/ristretto/musig.rs [hunk 5]
```diff
@@ -0,0 +1,960 @@
...
+    }
+
+    fn calculate_challenge<D: Digest>(
+        r_agg: &RistrettoPublicKey,
+        p_agg: &RistrettoPublicKey,
+        m: &MessageHash,
+    ) -> RistrettoSecretKey
+    {
+        let e = Challenge::<D>::new().concat(r_agg.to_bytes()).concat(p_agg.to_bytes()).concat(m).hash();
+        RistrettoSecretKey::from_vec(&e).expect("Found a u256 that does not map to a valid Ristretto scalar")
```
[the remark this answers is not in the document]
Comment (CjS77) [reply infrastructure/crypto/src/ristretto/musig.rs]: In general, I agree.
In this case, `RistrettoSecreyKey::from_vec` should _never_ fail, so if it does, I'd be panicking!

[the remark this answers is not in the document]
Comment (philipr-za) [reply infrastructure/crypto/src/ristretto/musig.rs]: Hehehe, fair enough. Agree that there are cases where if fundamental components are not working that the system should rather panic.

File: infrastructure/crypto/src/ristretto/musig.rs [hunk 6]
```diff
@@ -0,0 +1,960 @@
...
+        let e = &self.challenge;
+        signature.verify(&p, e)
+    }
+
+    fn calculate_agg_signature(&self) -> RistrettoSchnorr {
+        self.partial_signatures.sum().unwrap()
+    }
+
+    fn set_signature<D: Digest>(mut self, index: usize, signature: RistrettoSchnorr) -> MuSigState {
+        if self.partial_signatures.set_item(index, signature) {
```
[the remark this answers is not in the document]
Comment (CjS77) [reply infrastructure/crypto/src/ristretto/musig.rs]: Agreed 👍

File: infrastructure/crypto/src/ristretto/musig.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (CjS77) [reply infrastructure/crypto/src/ristretto/musig.rs]: I went with `set_message` taking an arbitrary byte string, but immediately hashing it and storing the hashed message

**Pull request merged**